components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/iwa/ntlm/NTLMAuthenticationGrantHandlerWithHandshake.java

```diff
@@ -0,0 +1,103 @@
+    private static final String securityPackage = "Negotiate";
+
+    public int getNLTMMessageType(byte[] decodedNLTMMessage) throws IdentityOAuth2Exception {
+        int messageType;
+        if (decodedNLTMMessage.length > 8) {
+            messageType = decodedNLTMMessage[8];
+        } else {
+            throw new IdentityOAuth2Exception(
+                    "Cannot extract message type from NLTM Token. Decoded token length is less than 8.");
+        }
+
+        //NLTM token type must be one of 1,2 or 3
+        if (messageType < 1 || messageType > 3) {
+            throw new IdentityOAuth2Exception(
+                    "Invalid NLTM message type:" + messageType + ". Should be one of 1,2 or 3.");
+        }
+
+        return messageType;
+    }
+
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+        boolean validGrant = super.validateGrant(tokReqMsgCtx);
+
+        if (!validGrant) {
+            return false;
+        }
+
+        String token = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getWindowsToken();
+
+        IWindowsSecurityContext serverContext = null;
+        if (token != null) {
+
+            byte[] bytesToken = Base64.decode(token);
+            int tokenType = getNLTMMessageType(bytesToken);
+
+            if (log.isDebugEnabled()) {
+                log.debug("Received NTLM token Type " + tokenType + ":" + token);
+            }
+
+            if (tokenType == 1) {
+                serverContext = provider.acceptSecurityToken("server-connection", bytesToken, securityPackage);
+                String type2Token = Base64.encode(serverContext.getToken());
+                if (log.isDebugEnabled()) {
+                    log.debug("Sent NTLM token Type 2:" + type2Token);
+                }
+                ResponseHeader[] responseHeaders = new ResponseHeader[1];
+                responseHeaders[0] = new ResponseHeader();
+                responseHeaders[0].setKey("WWW-Authenticate");
+                responseHeaders[0].setValue("NTLM " + type2Token);
+                tokReqMsgCtx.addProperty("RESPONSE_HEADERS_PROPERTY", responseHeaders);
+                return false;
+            } else if (tokenType == 3) {
+                serverContext = provider.acceptSecurityToken("server-connection", bytesToken, securityPackage);
+                String resourceOwnerUserNameWithDomain = serverContext.getIdentity().getFqn();
+                String resourceOwnerUserName = resourceOwnerUserNameWithDomain.split("\\\\")[1];
+                tokReqMsgCtx.setAuthorizedUser(OAuth2Util.getUserFromUserName(resourceOwnerUserName));
+                return true;
+            } else {
+                log.error("Unknown NTLM token, Type " + tokenType + ":" + token);
+                return false;
+            }
+
+        } else {
+            log.error("Received NTLM token is null");
```

Comment:
Do we have to log the message here since we are throwing exception with same message